timemachine/md/hrex.py

```diff
@@ -0,0 +1,247 @@
+from dataclasses import dataclass
+from typing import Callable, Generic, List, NewType, Optional, Sequence, Tuple, TypeVar
+
+import numpy as np
+from numpy.typing import NDArray
+
+from timemachine.md.moves import MetropolisHastingsMove, MixtureOfMoves
+from timemachine.utils import batches
+
+_Replica = TypeVar("_Replica")
+
+StateIdx = NewType("StateIdx", int)
+ReplicaIdx = NewType("ReplicaIdx", int)
+
+
+@dataclass
+class NeighborSwapMove(MetropolisHastingsMove[List[_Replica]]):
+    def __init__(self, log_q: Callable[[_Replica, StateIdx], float], s_a: StateIdx, s_b: StateIdx):
+        super().__init__()
+        self.log_q = log_q
+        self.s_a = s_a
+        self.s_b = s_b
+
+    def propose_with_log_q_diff(self, state: List[_Replica]) -> Tuple[List[_Replica], float]:
+        s_a = self.s_a
+        s_b = self.s_b
+        state_ = list(state)
+        state_[s_a], state_[s_b] = state_[s_b], state_[s_a]
```

Comment:
Would be cleaner if not modifying the state of the object that state is coming from. This assumes that this swap is not sequential (load values s_b and s_a before assigning the values) which I assume it does, but non-obvious reading the code.

```suggestion
        state_[s_a], state_[s_b] = state[s_b], state[s_a]
```

Reply:
Swap assignments in Python are "simultaneous", e.g.

```python
>>> x = 1
>>> y = 2
>>> x, y = y, x
>>> x, y
(2, 1)
```
but will apply the change to disambiguate.